param/parameterized.py

```diff
@@ -3573,39 +3574,65 @@ def _name_if_set(parameterized):
     return '' if default_name else parameterized.name
 
 
-def _get_param_repr(key, val, p, truncate=40):
+def truncate(str_, maxlen = 30):
+    """Return HTML-safe truncated version of given string"""
+    rep = (str_[:(maxlen-2)] + '..') if (len(str_) > (maxlen-2)) else str_
+    return html.escape(rep)
+
+
+def _get_param_repr(key, val, p, vallen=30, doclen=40):
     """HTML representation for a single Parameter object and its value"""
     if hasattr(val, "_repr_html_"):
         try:
             value = val._repr_html_(open=False)
         except:
             value = val._repr_html_()
     else:
-        rep = repr(val)
-        value = (rep[:truncate] + '..') if len(rep) > truncate else rep
+        value = truncate(repr(val), vallen)
 
-    modes = []
-    if p.constant:
-        modes.append('constant')
-    if p.readonly:
-        modes.append('read-only')
-    if getattr(p, 'allow_None', False):
-        modes.append('nullable')
-    mode = ' | '.join(modes)
     if hasattr(p, 'bounds'):
-        bounds = p.bounds
+        if p.bounds is None:
+            range_ = '(-&infin;,&infin;)'
+        elif hasattr(p,'inclusive_bounds'):
+            # Numeric bounds use ( and [ to indicate exclusive and inclusive
+            bl,bu = p.bounds
+            il,iu = p.inclusive_bounds
+            lb = ('[' if il else '(') + ('-&infin;' if bl is None else str(bl))
+            ub = ('&infin;' if bu is None else str(bu)) + (']' if iu else ')')
+            range_ = lb + ', ' + ub
+        else:
+            range_ = repr(p.bounds)
     elif hasattr(p, 'objects') and p.objects:
-        bounds = ', '.join(list(map(repr, p.objects)))
+        range_ = ', '.join(list(map(repr, p.objects)))
+    elif hasattr(p, 'class_'):
+        if isinstance(p.class_, tuple):
+            range_ = ' | '.join(kls.__name__ for kls in p.class_)
+        else:
+            range_ = p.class_.__name__
+    elif hasattr(p, 'regex'):
+        range_ = '.*' if p.regex is None else str(p.regex)
```

Comment:
I don't think it's super frequent to define a regex for a `String` Parameter and wouldn't expect all Param users to be knowledgeable about regular expressions. I'd suggest either not including `'.*'` or being more explicit with e.g. `regex(.*)`.

Reply:
The reason to include it is only that when `allow_None` is `True` and there's no `regex`, the range is shown as `| None`, which I think is confusing, given that I read `|` as "or", so it's "???? or None" (i.e., "_what_ or None??"). Can you think of a better way to convey "any string" than `.*`? `''` would be accurate since the empty string is a regex that matches every string, and then it would show `'' | None`, if that would look better.  Or, sure, `regex(.*)`, if you think that's clearer.

Reply:
Hmm, I agree that most options here aren't very good. Not sure I love it but if there's no regex I'd also be okay with `str | None` if `allow_None` or `str` otherwise.

Reply:
Actually `str | None` for no regex + allow_None and nothing at all for no regex and `allow_None=False` is my preference.

Reply:
(That's the same reason I went with (-Inf,Inf) for numbers; it looked odd without it. If anyone has a solution that conveys "any valid input, or None" more clearly, happy to use that instead!)

Reply:
Hmm; looks like @philippjfr 's notes weren't updated on my screen until after I made the above note. I don't mind "str", but I'd like us to use the same approach for number parameters as for string, since it's precisely the same concept: any allowed string, or any allowed number.  Instead of (-Inf, Inf), would a number be "num" in this approach, as in "num | None"?

Reply:
Actually, None is a different type, so I suppose we really ought to be putting it into the Type column!  I.e. the type isn't really `Integer`, it's `Integer | None`! I think that will eliminate all the awkwardness on the range field from `| None`, so I'll change it to do that unless someone strongly objects. Seems like the obvious next step from Philipp's proposal.

Assuming we do that, then there is still a question whether to combine the type and range columns. Combining them makes it clear that the constraints apply to the non-None type, not None: `Integer | None`, `Integer [0, Inf) |  None`, `List<Integer> len(0,10) | None`.

I can't quite decide which I prefer; any votes?